consensus/state.go

```diff
@@ -522,33 +834,36 @@ func (cs *ConsensusState) createProposalBlock() (block *types.Block, blockParts
 // Enter: any +2/3 prevotes for future round.
 // Prevote for LockedBlock if we're locked, or ProposalBlock if valid.
 // Otherwise vote nil.
-func (cs *ConsensusState) EnterPrevote(height int, round int, timedOut bool) {
-	cs.mtx.Lock()
-	defer cs.mtx.Unlock()
+func (cs *ConsensusState) EnterPrevote(height int, round int) {
+	//cs.mtx.Lock()
+	//defer cs.mtx.Unlock()
 	if cs.Height != height || round < cs.Round || (cs.Round == round && RoundStepPrevote <= cs.Step) {
 		log.Debug(Fmt("EnterPrevote(%v/%v): Invalid args. Current step: %v/%v/%v", height, round, cs.Height, cs.Round, cs.Step))
 		return
 	}
 
 	// fire event for how we got here
-	if timedOut {
-		cs.evsw.FireEvent(types.EventStringTimeoutPropose(), cs.RoundStateEvent())
-	} else if cs.isProposalComplete() {
+	if cs.isProposalComplete() {
 		cs.evsw.FireEvent(types.EventStringCompleteProposal(), cs.RoundStateEvent())
 	} else {
 		// we received +2/3 prevotes for a future round
 		// TODO: catchup event?
 	}
 
+	// cs.stopTimer()
+
 	log.Info(Fmt("EnterPrevote(%v/%v). Current: %v/%v/%v", height, round, cs.Height, cs.Round, cs.Step))
 
+	// needs to happen before doPrevote because
+	// addVote will take us to the next step
+	// TODO: should we fire our own votes back through the receive Routine?
+	cs.updateRoundStep(round, RoundStepPrevote)
```

Comment:
Looks like you changed signAddVote to call sendInternalMessage, so cs.updateRoundStep(round, RoundStepPrevote) can go back below the doPrevote() call